### DISCUSSION TITLE
Let's test Node.js package

### DISCUSSION THREAD

**Discussion by thunder-coding** [termux/termux-packages#7758] [General]
Currently we don't test nodejs package because it's too big to test and cross compiled binaries can't be tested on CI. Let's fix this

1. Make sure you are having latest version of `nodejs`

```sh
pkg up
pkg I nodejs
```

2. Download and extract source tarball corresponding to your nodejs version

```sh
export NODE_VERSION=$(node --version)
curl https://nodejs.org/dist/$NODE_VERSION/node$NODE_VERSION-.tar.gz -Lo node-$NODE_VERSION.tar.gz
tar --extract node-$NODE_VERSION.tar.gz
cd node-$NODE_VERSION
```

3. Fake compiled node

Note: This is highly recommended against building node locally since building from source takes about 3 hours even on CI

```sh
mkdir -p out/Release
ln -s $PREFIX/bin/node out/Release/node
```

4. Run the tests

```sh
./tools/test.py -p verbose > test-result.txt
```

5. Share here list of failed tests with output of your `termux-info`

For list of failed test, all you need to do is
```sh
grep FAIL test-result.txt | termux-clipboard-set
```

Known list of failing tests

- [x] test-child-process-exec-env (Hard-coded path)
- [x] test-child-process-spawnsync-shell (Hard-coded path)
- [x] test-child-process-stdio-overlapped (Requires a binary compiled during nodejs build process to be available in path)
- [ ] test-child-process-uid-gid
- [x] test-datetime-change-notify (#7776)
- [ ] test-dgram-bind-fd
- [ ] test-dgram-membership
- [ ] test-dgram-socket-buffer-size
- [ ] test-dgram-udp6-link-local-address
- [ ] test-fs-link
- [ ] test-fs-promises
- [ ] test-fs-readfile-error
- [ ] test-gc-http-client-onerror
- [ ] test-os
- [ ] test-process-constants-noatime
- [x] test-process-env-tz (#7776)
- [ ] test-trace-events-fs-sync
- [ ] test-report-fatal-error
- [ ] test-report-getreport
- [ ] test-report-signal
- [ ] test-report-uncaught-exception
- [ ] test-report-uncaught-exception-compat
- [ ] test-report-uncaught-exception-primitives
- [ ] test-report-uncaught-exception-symbols
- [ ] test-report-uv-handles
- [ ] test-report-worker
- [ ] test-report-writereport
- [ ] test-wasi-symlinks
- [ ] test-fs-watch

**Comment by thunder-coding** [comment]
For `test-child-process-uid-gid`, this seems too be an Android problem. `setuid()` and `setgid()` don't do anything and they neither give any error.
 https://stackoverflow.com/questions/39893422/aosp-setgid-setuid-have-no-effect

This can be fixed on Node.js' side by patching `lib/child_process.js` to return an error on passage of `{uid:0}` and `{gid:0}` when the current user isn't root

**Reply by ghost** [reply]
> setuid() and setgid() don't do anything and they neither give any error.

They require binary to have modes `6755`, `4755` (setuid only) or `2755` (setgid) and ownership `root:root`, which isn't possible without root permissions and disabled `nosuid` mount option for /data partition. Moreover we tend to patch these calls as they lead to process being terminated by seccomp.

Termux doesn't support setuid/setgid binaries officially.

**Comment by thunder-coding** [comment]
About `test-datetime-change-notify` and `test-process-env-tz`, It seems to be a problem with libicu.

https://github.com/unicode-org/icu/blob/0e7b4428866f3133b4abba2d932ee3faa708db1d/icu4c/source/common/putil.cpp#L1135-1138


For Android, libicu doesn't seem to respect `TZ` environment variable. Here is the commit which caused this change https://github.com/unicode-org/icu/commit/c9037ca8d3c0738f1a044dd845bdc0dc646f3e97

**Reply by thunder-coding** [reply]
Fixed by #7776
